### PR TITLE
Measurement tampering

### PR DESCRIPTION
**Pull request by oliveradk**
(all tests passing)

- language model support (from language-models branch)
- multi-label classification support 
- measurment tampering predictor training and dataset support
- misc bug fixees

### CODE REVIEW

File: src/cupbearer/data/tampering.py [hunk 1]
```diff
@@ -0,0 +1,62 @@
+from dataclasses import dataclass
+from typing import ClassVar
+
+import torch
+from datasets import load_dataset
+
+from . import DatasetConfig
+
+
+class TamperingDataset(torch.utils.data.Dataset):
```
Comment (ejnnr) [inline src/cupbearer/data/tampering.py:10]: To define a measurement tampering task, I see two approaches (with the task format from #30):
1. Add options to `TamperingDataset` to only return a selection of the data (e.g. only easy training data, or only validation data that has tampering). Then use `Task.from_separate_data()` to recombine them. Advantage is we get some control over mixing ratio and use the standard interface, downside is that it's kind of indirect and complicated.
2. Pass `trusted_data`, `untrusted_train_data`, and `test_data` directly to `Task`, without using `MixedData` at all. This is closer to the format the datasets are already in anyway, so probably simpler. We'll just have to make sure to return the right label format: `trusted_data` and `untrusted_train_data` should just return `(text, measurements)`, whereas `test_data` should return `((text, measurements), is_clean)` I think.

Leaning a bit towards 2. overall but open to either approach or some third one

Comment (ejnnr) [reply src/cupbearer/data/tampering.py:10]: To be clear, no need to do this before merging, just noting down my thoughts

File: src/cupbearer/models/transformers_hf.py [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+        self.model = model
+        self.tokenizer = tokenizer
+        self.max_length = max_length
+        self.embed_dim = embed_dim
+
+        # setup
+        self.tokenizer.pad_token = tokenizer.eos_token
+
+    @property
+    def defeault_names(self) -> list[str]:
```
Comment (ejnnr) [inline src/cupbearer/models/transformers_hf.py:78]:
```suggestion
    def default_names(self) -> list[str]:
```
Typo

File: src/cupbearer/models/transformers_hf.py [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+        self.num_classes = num_classes
+        self.classifier = nn.Linear(self.embed_dim, self.num_classes)
+
+    def forward(self, x: str | list[str]):
+        # get tokens
+        tokens = self.process_input(x)
+        # get embeddings
+        embeddings = self.get_embeddings(tokens)
+
+        # TODO (odk) se store (doesn't this slow down training?)
```
Comment (ejnnr) [inline src/cupbearer/models/transformers_hf.py:124]: > (doesn't this slow down training?)

Calling `self.store()` should be ~free outside a `self.capture()` context manager

Comment (ejnnr) [reply src/cupbearer/models/transformers_hf.py:124]: (but also like we discussed, we should switch to pytorch hooks anyway, so it won't be each model's responsibility anymore to expose activations)

File: src/cupbearer/models/transformers_hf.py [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+
+        # sensor embeddings
+        batch_inds, seq_ids = torch.where(
+            tokens["input_ids"] == self.sensor_token_id
+        )  # TODO: pre-specify that its always 3?
+        sensor_embs = embeddings[batch_inds, seq_ids].reshape(
+            b, self.n_sensors, self.embed_dim
+        )
+        # last token embedding (for aggregate measurement)
+        last_token_ind = tokens["attention_mask"].sum(dim=1) - 1
```
Comment (ejnnr) [inline src/cupbearer/models/transformers_hf.py:176]: We should probably add a check somewhere that padding is on the right just to avoid nasty surprises

File: src/cupbearer/models/transformers_hf.py [hunk 5]
```diff
@@ -0,0 +1,189 @@
...
+        last_embs = embeddings[torch.arange(b), last_token_ind]
+        probe_embs = torch.concat([sensor_embs, last_embs.unsqueeze(dim=1)], axis=1)
+        assert probe_embs.shape == (b, self.n_probes, self.embed_dim)
+        logits = torch.concat(
```
Comment (ejnnr) [inline src/cupbearer/models/transformers_hf.py:180]: Probably simpler to make `self.probes` a single `nn.Linear` with `n_probes` output logits? Though if we ever have non-linear probes, then it actually makes a difference (do we share most of their body or not between measurements), not sure which option we'd want in that case

File: src/cupbearer/scripts/_shared.py [hunk 6]
```diff
@@ -32,34 +39,43 @@ def __init__(
...
         self.test_accuracy = torch.nn.ModuleList(
             [
-                Accuracy(task="multiclass", num_classes=num_classes)
+                Accuracy(task=self.task, num_classes=num_classes, num_labels=num_labels)
                 for _ in test_loader_names
             ]
         )
 
+    def _get_loss_func(self, task):
+        if self.task == torch_enums.ClassificationTask.MULTICLASS:
```
Comment (ejnnr) [inline src/cupbearer/scripts/_shared.py:71]:
```suggestion
        if self.task == "multiclass":
```
I think it would also be reasonable to use the `torchmetrics` enum everywhere (i.e. in the type annotation and default argument), but we should probably be consistent (and if we do use the enum, make sure we support all its possible values of course)

**Review by ejnnr — CHANGES_REQUESTED** [review]
Left a few minor comments but mostly looks good for now! (of course there's stuff to finish up later) In addition to comments, I don't think `demo.ipynb` has any changes that should be committed? `measurement tampering.ipynb` also seems partially outdated and like we probably don't need it in the repository, but I do like the dataset exploration in the beginning, we can reuse that when we make a nicer demo notebook for measurement tampering at some point

**Comment by ejnnr** [comment]
@oliveradk I merged it after making a few small changes (you might want to double check them). The remaining open comments seem less important, just take a look at them at some point
